towerlib/towerlib.py

```diff
@@ -1997,24 +2011,579 @@ def schedules(self):
+                                               .format(project.name, credential_id))
+                            return credential_id
+        return credential_id
+
+    def get_projects_by_scm_url_and_branch(self, scm_url, scm_branch):
+        """Get the ansible tower projects with given scm_url and branch name.
+
+        Args:
+            scm_url: given scm url.
+            scm_branch: given scm branch.
+
+        Returns:
+            list: list of matching projects
+
+        """
+        matching_projects = []
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_type == 'git':
+                if project.scm_url == scm_url and project.scm_branch == scm_branch:
+                    matching_projects.append(project)
+        return matching_projects
+
+    def get_projects_by_scm_url(self, scm_url):
+        """Get the ansible tower projects with given scm_url.
+
+        Args:
+            scm_url: the scm url of the production branch project.
+
+        Returns:
+            The id of the production branch project.
+
+        """
+        matching_projects = []
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_type == 'git':
+                if project.scm_url == scm_url:
+                    matching_projects.append(project)
+        return matching_projects
+
+    def get_playbooks_by_scm_url_and_branch(self, scm_url, scm_branch):
+        """Get all the playbooks of the project with the given scm_url and branch name.
+
+        Args:
+            scm_url: the given scm url.
+            scm_branch: the given scm branch
+
+        Returns:
+            list: a list with the names of all the playbooks for the matching project.
+
+        """
+        playbooks = None
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_url == scm_url and project.scm_branch == scm_branch:
+                playbooks = project.playbooks
+                break
+        return playbooks
+
+    def get_production_branch_playbooks(self, scm_url):
+        """Get all the playbooks of the project, which has the branch production as scm branch.
+
+        Args:
+            scm_url: given scm url
+
+        Returns:
+            list: a list with the names of all the playbooks of production branch project.
+
+        """
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_url == scm_url and project.scm_branch == 'production':
+                playbooks = project.playbooks
+                return playbooks
+
+    def get_ansible_facts_by_host_id(self, host_id):
+        """Get the ansible_facts of the given host.
+
+        Args:
+            host_id: id of the host for which the method will return the ansible_facts.
+
+        Returns:
+            dict: ansible_facts as dictionary.
+
+        """
+        host = self.get_host_by_id(host_id)
```

Comment:
This should be made safe. It is not sure that the host_id exists and if not this code would raise an exception for an attribute error.

Reply:
Done https://github.com/schubergphilis/towerlib/pull/79/commits/60bdc724156ec9b7c8519cdb68dc6eb6a2091815